<a href="https://colab.research.google.com/github/ajisakti554/highlight_project/blob/main/StudentPerformance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Academic Performance KPI**

Ini adalah KPI untuk melihat keberhasilan akademik secara langsung.

1. Average Exam Score (Rata-rata Skor Ujian): Mengukur performa rata-rata seluruh siswa.

2. Passing Rate (%): Persentase siswa yang mencapai skor di atas ambang batas tertentu (misalnya >70).

3. Improvement Index: Selisih antara Previous_Scores dengan Exam_Score. Ini penting untuk melihat apakah siswa mengalami progres atau penurunan.

**KPI 1**

# 1. Data Cleaning & Pre-processing

In [ ]:
import pandas as pd

In [ ]:
df=pd.read_csv('/content/StudentPerformanceFactors.csv')
df

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0                23          84                  Low                High   
1                19          64                  Low              Medium   
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
4                19          92               Medium              Medium   
...             ...         ...                  ...                 ...   
6602             25          69                 High              Medium   
6603             23          76                 High              Medium   
6604             20          90               Medium                 Low   
6605             10          86                 High                High   
6606             15          67               Medium                 Low   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
0                            No            7               73   
1                            No            8               59   
2                           Yes            7               91   
3                           Yes            8               98   
4                           Yes            6               65   
...                         ...          ...              ...   
6602                         No            7               76   
6603                         No            8               81   
6604                        Yes            6               65   
6605                        Yes            6               91   
6606                        Yes            9               94   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
0                 Low             Yes                  0           Low   
1                 Low             Yes                  2        Medium   
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
4              Medium             Yes                  3        Medium   
...               ...             ...                ...           ...   
6602           Medium             Yes                  1          High   
6603           Medium             Yes                  3           Low   
6604              Low             Yes                  3           Low   
6605             High             Yes                  2           Low   
6606           Medium             Yes                  0        Medium   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
0             Medium      Public       Positive                  3   
1             Medium      Public       Negative                  4   
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
4               High      Public        Neutral                  4   
...              ...         ...            ...                ...   
6602          Medium      Public       Positive                  2   
6603            High      Public       Positive                  2   
6604          Medium      Public       Negative                  2   
6605          Medium     Private       Positive                  3   
6606          Medium      Public       Positive                  4   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
0                       No              High School               Near   
1                       No                  College           Moderate   
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
4                       No                  College               Near   
...                    ...                      ...                ...   
6602                    No              High School             

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Hours_Studied               6607 non-null   int64 
 1   Attendance                  6607 non-null   int64 
 2   Parental_Involvement        6607 non-null   object
 3   Access_to_Resources         6607 non-null   object
 4   Extracurricular_Activities  6607 non-null   object
 5   Sleep_Hours                 6607 non-null   int64 
 6   Previous_Scores             6607 non-null   int64 
 7   Motivation_Level            6607 non-null   object
 8   Internet_Access             6607 non-null   object
 9   Tutoring_Sessions           6607 non-null   int64 
 10  Family_Income               6607 non-null   object
 11  Teacher_Quality             6529 non-null   object
 12  School_Type                 6607 non-null   object
 13  Peer_Influence              6607 non-null   obje

In [ ]:
df.isnull().sum()

Hours_Studied                  0
Attendance                     0
Parental_Involvement           0
Access_to_Resources            0
Extracurricular_Activities     0
Sleep_Hours                    0
Previous_Scores                0
Motivation_Level               0
Internet_Access                0
Tutoring_Sessions              0
Family_Income                  0
Teacher_Quality               78
School_Type                    0
Peer_Influence                 0
Physical_Activity              0
Learning_Disabilities          0
Parental_Education_Level      90
Distance_from_Home            67
Gender                         0
Exam_Score                     0
dtype: int64

In [ ]:
import pandas as pd
import numpy as np

# 1. Menangani missing values pada kolom kategorikal
cols_with_nan = ['Teacher_Quality', 'Parental_Education_Level', 'Distance_from_Home']
for col in cols_with_nan:
    df[col] = df[col].fillna(df[col].mode()[0])

# 2. Cek apakah ada skor yang di luar nalar (misal skor 0-100)
print(df[['Previous_Scores', 'Exam_Score']].describe())

# 3. Menghapus baris yang mungkin memiliki skor null (jika ada)
df.dropna(subset=['Exam_Score', 'Previous_Scores'], inplace=True)

       Previous_Scores   Exam_Score
count      6607.000000  6607.000000
mean         75.070531    67.235659
std          14.399784     3.890456
min          50.000000    55.000000
25%          63.000000    65.000000
50%          75.000000    67.000000
75%          88.000000    69.000000
max         100.000000   101.000000


Nilai exam yang >100 harus dihapus, setelah dicek, hanya ada data yang >100, yaitu 101 (delete solusinya)

In [ ]:
df.drop(df[df['Exam_Score'] > 100].index, inplace=True)

In [ ]:
df['Exam_Score'].value_counts()

Exam_Score
68     759
66     751
67     717
65     679
69     624
70     542
64     501
71     408
63     371
72     304
62     264
61     171
73     141
74     106
60      77
75      48
59      40
58      22
76      16
80       5
77       5
86       4
94       4
78       4
57       4
82       4
79       3
89       3
98       3
88       3
97       3
84       3
93       2
95       2
99       2
92       2
87       2
100      1
83       1
55       1
91       1
96       1
85       1
56       1
Name: count, dtype: int64

In [ ]:
df.duplicated().sum()

np.int64(0)

# 2.Feature Engineering (Membuat Rumus KPI)

In [ ]:

# 2. Membuat kolom 'Is_Pass' (Boolean: True jika >= 70)
# Ini untuk menghitung Passing Rate nanti
df['Is_Pass'] = df['Exam_Score'] >= 70

# 3. Membuat kolom 'Improvement_Score'
df['Improvement_Score'] = df['Exam_Score'] - df['Previous_Scores']

# 4. Membuat kategori performa untuk visualisasi nanti
def categorize_performance(score):
    if score >= 85: return 'Excellent'
    elif score >= 70: return 'Good'
    elif score >= 55: return 'Average'
    else: return 'Below Average'

df['Performance_Category'] = df['Exam_Score'].apply(categorize_performance)

# Cek 5 data teratas
print(df[['Previous_Scores', 'Exam_Score', 'Improvement_Score', 'Is_Pass', 'Performance_Category']].head())

   Previous_Scores  Exam_Score  Improvement_Score  Is_Pass  \
0               73          67                 -6    False   
1               59          61                  2    False   
2               91          74                -17     True   
3               98          71                -27     True   
4               65          70                  5     True   

  Performance_Category  
0              Average  
1              Average  
2                 Good  
3                 Good  
4                 Good  


# 3. EDA

In [ ]:
# 1. Menghitung Metrik Utama
overall_avg = df['Exam_Score'].mean()
passing_rate = df['Is_Pass'].mean() * 100
avg_improvement = df['Improvement_Score'].mean()

# 2. Menghitung Distribusi Kategori (untuk melihat sebaran kemampuan)
performance_dist = df['Performance_Category'].value_counts(normalize=True) * 100

# 3. Menampilkan Laporan KPI
print("========== ACADEMIC PERFORMANCE REPORT ==========")
print(f"Rata-rata Skor Ujian (Current) : {overall_avg:.2f}")
print(f"Rata-rata Skor Sebelumnya      : {df['Previous_Scores'].mean():.2f}")
print(f"Indeks Kemajuan (Improvement)  : {avg_improvement:.2f}")
print(f"Tingkat Kelulusan (Pass Rate)  : {passing_rate:.2f}%")
print("==================================================")
print("\nDistribusi Performa Siswa (%):")
print(performance_dist.round(2))

========== ACADEMIC PERFORMANCE REPORT ==========
Rata-rata Skor Ujian (Current) : 67.23
Rata-rata Skor Sebelumnya      : 75.07
Indeks Kemajuan (Improvement)  : -7.84
Tingkat Kelulusan (Pass Rate)  : 24.58%

Distribusi Performa Siswa (%):
Performance_Category
Average      75.42
Good         24.07
Excellent     0.51
Name: proportion, dtype: float64


1. Indikator Krisis: Penurunan Performa Kolektif
* Indeks Kemajuan (Improvement) -7.84: Ini adalah temuan paling kritikal. Secara rata-rata, nilai siswa turun hampir 8 poin dibandingkan ujian sebelumnya. Dalam skala pendidikan, penurunan rata-rata sebesar ini di 6.000+ siswa menunjukkan adanya masalah sistemik, bukan sekadar masalah individu.

* Gap Skor: Rata-rata turun dari 75.07 (masuk kategori Good) menjadi 67.23 (turun ke kategori Average).

2. Tingkat Kelulusan (Pass Rate) yang Mengkhawatirkan
* Hanya 24.58% siswa yang lulus: Dengan standar kelulusan (passing grade) 70 yang kita tetapkan, hanya 1 dari 4 siswa yang berhasil lulus.

* Ini adalah informasi kunci bagi pemangku kepentingan (stakeholder). Jika ini adalah sekolah nyata, kepala sekolah harus segera melakukan evaluasi kurikulum atau metode pengajaran.

3. Distribusi Performa: Penumpukan di "Average"
* 75.42% Siswa berada di kategori Average: Mayoritas besar siswa terjebak di level menengah bawah.

* Excellent hanya 0.51%: Hampir tidak ada siswa yang mencapai skor istimewa (>= 85). Ini menunjukkan bahwa sistem pendidikan saat ini mungkin hanya mampu membuat siswa "bertahan" tapi tidak mampu mendorong mereka untuk "unggul".

4. Analisis Statistik (Standard Deviation)
* Jika kamu perhatikan di gambar describe(), Std (Standard Deviation) untuk Exam_Score (3.89) jauh lebih kecil daripada Previous_Scores (14.39).

* Artinya: Dulu nilai siswa sangat bervariasi (ada yang sangat pintar, ada yang sangat kurang). Sekarang, nilai semua siswa menjadi seragam tapi rendah. Nilai mereka "mengumpul" di angka 67.

In [ ]:
# Segmentasi Pass Rate berdasarkan Akses Internet
internet_analysis = df.groupby('Internet_Access').agg(
    Avg_Score=('Exam_Score', 'mean'),
    Pass_Rate=('Is_Pass', lambda x: x.mean() * 100)
).round(2)

# Segmentasi Pass Rate berdasarkan Pendapatan Keluarga
income_analysis = df.groupby('Family_Income').agg(
    Avg_Score=('Exam_Score', 'mean'),
    Pass_Rate=('Is_Pass', lambda x: x.mean() * 100)
).round(2)

print("--- Analisis Berdasarkan Akses Internet ---")
print(internet_analysis)
print("\n--- Analisis Berdasarkan Pendapatan Keluarga ---")
print(income_analysis)

--- Analisis Berdasarkan Akses Internet ---
                 Avg_Score  Pass_Rate
Internet_Access                      
No                   66.47      17.47
Yes                  67.29      25.16

--- Analisis Berdasarkan Pendapatan Keluarga ---
               Avg_Score  Pass_Rate
Family_Income                      
High               67.82      29.34
Low                66.85      21.15
Medium             67.33      25.77


1. Analisis Deskriptif: Fenomena "Penurunan Massal"
Berdasarkan statistik deskriptif di gambar pertama:

* Standar Deviasi (std) Menyusut: Pada Previous_Scores, std adalah 14.39, namun pada Exam_Score turun drastis menjadi 3.89.

* Artinya: Dulu nilai siswa sangat bervariasi (ada yang sangat pintar, ada yang sangat kurang). Sekarang, nilai semua siswa menjadi sangat seragam (homogen) tapi di level yang rendah. Mereka "berkumpul" di sekitar angka 67.

2. Segmentasi: Mencari "Akar Masalah" (Deep Dive)
Inilah bagian paling krusial yang kamu hasilkan di gambar terakhir. Kita mulai melihat adanya ketimpangan:

* A. Dampak Akses Internet
Siswa dengan internet memiliki Pass Rate 25.16%, sementara yang tidak punya hanya 17.47%.

* Insight: Ada selisih sekitar ~8% pada tingkat kelulusan. Ini menunjukkan bahwa infrastruktur digital merupakan faktor pendukung performa, namun bukan satu-satunya penentu karena rata-rata skor mereka (Avg_Score) tidak berbeda jauh (67.29 vs 66.47).

B. Faktor Ekonomi (Family Income)
* Ada korelasi linier yang jelas di sini:

* High Income: Pass Rate 29.34%

* Medium Income: Pass Rate 25.77%

* Low Income: Pass Rate 21.15%

* Insight: Siswa dari keluarga mampu memiliki peluang lulus ~40% lebih tinggi dibandingkan siswa dari keluarga kurang mampu. Sebagai analis, kita bisa menyimpulkan adanya kesenjangan ekonomi yang memengaruhi performa akademik.

# 4. SQL & Data Export

In [ ]:

# Kita membuat kolom 'Is_Pass' secara manual
df['Is_Pass'] = df['Exam_Score'] >= 70

# Verifikasi apakah kolom sudah muncul
print("Daftar Kolom Terbaru:", df.columns.tolist())
print(df[['Exam_Score', 'Is_Pass']].head())

Daftar Kolom Terbaru: ['Hours_Studied', 'Attendance', 'Parental_Involvement', 'Access_to_Resources', 'Extracurricular_Activities', 'Sleep_Hours', 'Previous_Scores', 'Motivation_Level', 'Internet_Access', 'Tutoring_Sessions', 'Family_Income', 'Teacher_Quality', 'School_Type', 'Peer_Influence', 'Physical_Activity', 'Learning_Disabilities', 'Parental_Education_Level', 'Distance_from_Home', 'Gender', 'Exam_Score', 'Is_Pass', 'Improvement_Score', 'Performance_Category']
   Exam_Score  Is_Pass
0          67    False
1          61    False
2          74     True
3          71     True
4          70     True


In [ ]:
df.head()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             23          84                  Low                High   
1             19          64                  Low              Medium   
2             24          98               Medium              Medium   
3             29          89                  Low              Medium   
4             19          92               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                         No            7               73              Low   
1                         No            8               59              Low   
2                        Yes            7               91           Medium   
3                        Yes            8               98           Medium   
4                        Yes            6               65           Medium   

  Internet_Access  Tutoring_Sessions  ... Peer_Influence Physical_Activity  \
0             Yes                  0  ...       Positive                 3   
1             Yes                  2  ...       Negative                 4   
2             Yes                  2  ...        Neutral                 4   
3             Yes                  1  ...       Negative                 4   
4             Yes                  3  ...        Neutral                 4   

  Learning_Disabilities Parental_Education_Level  Distance_from_Home  Gender  \
0                    No              High School                Near    Male   
1                    No                  College            Moderate  Female   
2                    No             Postgraduate                Near    Male   
3                    No              High School            Moderate    Male   
4                    No                  College                Near  Female   

  Exam_Score Is_Pass Improvement_Score  Performance_Category  
0         67   False                -6               Average  
1         61   False                 2               Average  
2         74    True               -17                  Good  
3         71    True               -27                  Good  
4         70    True                 5                  Good  

[5 rows x 23 columns]

In [ ]:
# 1. Pastikan outlier 101 sudah hilang (filter ulang untuk keamanan)
df_final = df[df['Exam_Score'] <= 100].copy()

# 2. Ekspor data lengkap untuk analisis detail di Tableau
df_final.to_csv('student_performance_cleaned.csv', index=False)

# 3. Membuat tabel agregasi untuk Dashboard Utama (Simulasi SQL View)
# Kita akan mengelompokkan berdasarkan faktor-faktor kunci
summary_table = df_final.groupby(['Internet_Access', 'Family_Income', 'School_Type']).agg(
    Total_Siswa=('Exam_Score', 'count'),
    Avg_Exam_Score=('Exam_Score', 'mean'),
    Passing_Rate=('Is_Pass', lambda x: x.mean() * 100)
).reset_index()

summary_table.to_csv('kpi_summary_performance.csv', index=False)

print("Data berhasil diekspor! Siap untuk tahap Tableau.")

Data berhasil diekspor! Siap untuk tahap Tableau.


In [ ]:
len(df)

6606

**KPI 2 At-Risk Students**

**Data already Clean**

# **At-Risk Students**

In [ ]:
#data summary
df[(df['Exam_Score'] < 60) & (df['Improvement_Score'] < 0)].describe()

Hours_Studied  Attendance  Sleep_Hours  Previous_Scores  \
count      48.000000   48.000000    48.000000        48.000000   
mean        9.687500   63.770833     7.041667        74.062500   
std         4.727269    3.931540     1.529265         9.731297   
min         2.000000   60.000000     4.000000        62.000000   
25%         5.000000   61.000000     6.000000        67.000000   
50%        10.000000   63.000000     7.000000        71.000000   
75%        13.000000   65.250000     8.000000        78.000000   
max        21.000000   79.000000    10.000000        98.000000   

       Tutoring_Sessions  Physical_Activity  Exam_Score  Improvement_Score  
count          48.000000          48.000000   48.000000          48.000000  
mean            0.979167           2.500000   58.375000         -15.687500  
std             1.157576           1.071884    0.890255           9.598662  
min             0.000000           0.000000   55.000000         -39.000000  
25%             0.000000           2.000000   58.000000         -20.250000  
50%             1.000000           2.000000   59.000000         -12.500000  
75%             1.000000           3.000000   59.000000          -9.000000  
max             4.000000           4.000000   59.000000          -3.000000

In [ ]:
# Mengubah ambang batas ke 60 (karena nilai min adalah 55.00)
at_risk_list = df[(df['Exam_Score'] < 60) & (df['Improvement_Score'] < 0)]

print(f"Jumlah Siswa Berisiko (Skor < 60): {len(at_risk_list)}")
print(f"At-Risk Rate: {(len(at_risk_list) / len(df)) * 100:.2f}%")

Jumlah Siswa Berisiko (Skor < 60): 48
At-Risk Rate: 0.73%


In [ ]:
# 1. Mengambil daftar 48 siswa berisiko
at_risk_list = df[(df['Exam_Score'] < 60) & ((df['Exam_Score'] - df['Previous_Scores']) < 0)]

# 2. Menyimpan daftar tersebut ke file Excel secara permanen
at_risk_list.to_excel('Daftar_Siswa_Berisiko_Final.xlsx', index=True)

print("File berhasil dibuat! Silakan cek folder file di sebelah kiri (ikon folder) dan download file tersebut.")

File berhasil dibuat! Silakan cek folder file di sebelah kiri (ikon folder) dan download file tersebut.
